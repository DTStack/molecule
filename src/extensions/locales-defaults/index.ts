import { IExtension, IContributeType } from 'mo/model/extension';

const koKR = require('./locales/ko-KR.json');
const zhCN = require('./locales/zh-CN.json');
const en = require('./locales/en.json');
const locales = [zhCN, en, koKR];

export const ExtendsLocales: IExtension = {
    id: 'ExtendsLocales',
    name: 'Extends locales',
    contributes: {
        [IContributeType.Languages]: locales,
    },
    activate() {},
    dispose() {},
};
export const BuiltInLocales = locales;
export const BuiltInId = en.id;
export const BuiltInDefault = locales.find((item) => item.id === BuiltInId);
