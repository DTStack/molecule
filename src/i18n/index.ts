export { LocaleService } from './localeService';
export type { ILocaleService } from './localeService';
export type { ILocale } from './localization';
export type { ILocalizeProps } from './localize';
export { localize } from './localize';
