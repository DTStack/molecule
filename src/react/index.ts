export * from './component';
export * from './helper';
export * from './connector';
export * from './controller';
