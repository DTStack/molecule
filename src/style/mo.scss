@import 'mo/style/common';
@import 'mo/style/animation';

// ============= The Workbench parts =============
@import 'mo/components/actionBar/style';
@import 'mo/components/breadcrumb/style';
@import 'mo/components/button/style';
@import 'mo/components/checkbox/style';
@import 'mo/components/collapse/style';
@import 'mo/components/contextView/style';
@import 'mo/components/dialog/style';
@import 'mo/components/dropdown/style';
@import 'mo/components/input/style';
@import 'mo/components/list/style';
@import 'mo/components/menu/style';
@import 'mo/components/scrollable/style';
@import 'mo/components/select/style';
@import 'mo/components/tabs/style';
@import 'mo/components/toolbar/style';
@import 'mo/components/tree/style';
@import 'mo/components/search/style';
@import 'mo/components/tooltip/style';
@import 'mo/components/split/style';

// ============= The Workbench core component =============
@import 'mo/workbench/style';
@import 'mo/workbench/activityBar/style';
@import 'mo/workbench/editor/style';
@import 'mo/workbench/editor/welcome/style';
@import 'mo/workbench/menuBar/style';
@import 'mo/workbench/panel/style';
@import 'mo/workbench/settings/style';
@import 'mo/workbench/sidebar/style';
@import 'mo/workbench/statusBar/style';
@import 'mo/workbench/notification/style';
@import 'mo/workbench/sidebar/explore/style';
@import 'mo/workbench/problems/style';
@import 'mo/workbench/notification/statusBarView/style';

.#{$prefix} {
    bottom: 0;
    box-sizing: border-box;
    font-size: 13px;
    height: 100%;
    left: 0;
    margin: 0;
    outline: 0;
    overflow: hidden;
    padding: 0;
    position: absolute;
    right: 0;
    top: 0;
    user-select: none;
    width: 100%;
}

.mac {
    font-family: -apple-system, BlinkMacSystemFont, system-ui, sans-serif;
}

.center {
    text-align: center;
}

.hidden {
    display: none;
}
