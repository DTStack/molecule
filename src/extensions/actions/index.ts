import { IBottomActivityBarItem, IContributeType, type IExtension, IMenuItemProps, UniqueId } from 'mo/types';
import { concatMenu } from 'mo/utils';

import AddCursorAboveAction from './addCursorAbove';
import AddCursorBelowAction from './addCursorBelow';
import AddSelectionToNextFindMatchAction from './addSelectionToNextFindMatchAction';
import CopyAction from './copy';
import CopyLineDownAction from './copyLineDown';
import CopyLineUpAction from './copyLineUp';
import CutAction from './cut';
import ExpandSelectionAction from './expandSelection';
import FindAction from './find';
import MoveLineDownAction from './moveLineDown';
import MoveLineUpAction from './moveLineUp';
import PasteAction from './paste';
import { QuickAccessCommandAction } from './quickAccessCommandAction';
import { QuickAccessSettingsAction } from './quickAccessSettingsAction';
import QuickJumpToLineAction from './quickJumpToLineAction';
import QuickSelectLocaleAction from './quickSelectLocaleAction';
import QuickSelectThemeAction from './quickSelectThemeAction';
import QuickTogglePanelAction from './quickTogglePanelAction';
import QuickToggleSidebarAction from './quickToggleSideBarAction';
import RedoAction from './redo';
import ReplaceAction from './replace';
import SelectAllAction from './selectAll';
import SelectHighlightsAction from './selectHighlightsAction';
import ShrinkSelectionAction from './shrinkSelection';
import ToggleBlockCommentAction from './toggleBlockComment';
import ToggleLineCommentAction from './toggleLineComment';
import UndoAction from './undo';

export const ExtendsActions: IExtension = {
    id: 'ExtendsActions',
    name: 'Extend The Default Actions',
    contributes: {
        [IContributeType.Commands]: [
            UndoAction,
            RedoAction,
            SelectAllAction,
            CopyLineUpAction,
            CopyLineDownAction,
            MoveLineUpAction,
            MoveLineDownAction,
            CutAction,
            CopyAction,
            PasteAction,
            FindAction,
            ReplaceAction,
            AddCursorAboveAction,
            AddCursorBelowAction,
            ExpandSelectionAction,
            ShrinkSelectionAction,
            ToggleLineCommentAction,
            ToggleBlockCommentAction,
            AddSelectionToNextFindMatchAction,
            SelectHighlightsAction,
            QuickSelectThemeAction,
            QuickTogglePanelAction,
            QuickToggleSidebarAction,
            QuickJumpToLineAction,
            QuickAccessSettingsAction,
            QuickAccessCommandAction,
            QuickSelectLocaleAction,
        ],
    },
    activate: function (molecule): void {
        // append actions into settings' menu
        appendActionToSettingMenu(QuickAccessCommandAction);
        appendActionToSettingMenu(QuickAccessSettingsAction);
        appendActionToSettingMenu(QuickSelectThemeAction);

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_EDIT)
            .with(CutAction)
            .with(CopyAction)
            .with(PasteAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_EDIT)
            .with(FindAction)
            .with(ReplaceAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_EDIT)
            .with(ToggleLineCommentAction)
            .with(ToggleBlockCommentAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_SELECTION)
            .with(SelectAllAction)
            .with(ExpandSelectionAction)
            .with(ShrinkSelectionAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_SELECTION)
            .with(CopyLineUpAction)
            .with(CopyLineDownAction)
            .with(MoveLineUpAction)
            .with(MoveLineDownAction)
            .exhaust();

        appendActionGroupBy(molecule.builtin.getConstants().MENUBAR_ITEM_SELECTION)
            .with(AddCursorAboveAction)
            .with(AddCursorBelowAction)
            .with(AddSelectionToNextFindMatchAction)
            .with(SelectHighlightsAction)
            .exhaust();

        // update menu's keybinding
        updateMenuKeybinding(QuickAccessCommandAction.ID);
        updateMenuKeybinding(QuickTogglePanelAction.ID);
        updateMenuKeybinding(QuickToggleSidebarAction.ID);
        updateMenuKeybinding(UndoAction.ID);
        updateMenuKeybinding(RedoAction.ID);
        updateMenuKeybinding(CutAction.ID);

        function appendActionToSettingMenu(ctor: { ID: string }) {
            const keybinding = molecule.action.queryGlobalKeybinding(ctor.ID);
            // Add Settings into setting's menus
            molecule.activityBar.update<IBottomActivityBarItem>(
                molecule.builtin.getState().constants.ACTIVITYBAR_ITEM_SETTING,
                (prev) => ({
                    id: prev.id,
                    contextMenu: [
                        ...(prev.contextMenu || []),
                        {
                            id: ctor.ID,
                            name: molecule.locale.localize(ctor.ID, ctor.ID),
                            keybinding: keybinding
                                ? molecule.action.convertSimpleKeybindingToString(keybinding)
                                : undefined,
                        },
                    ],
                })
            );

            molecule.activityBar.onContextMenuClick((item) => {
                if (item.id === ctor.ID) {
                    molecule.action.execute(ctor.ID);
                }
            });
        }

        function updateMenuKeybinding(id: UniqueId) {
            const keybinding = molecule.action.queryGlobalKeybinding(id);
            if (keybinding) {
                molecule.menuBar.update({
                    id,
                    keybinding: molecule.action.convertSimpleKeybindingToString(keybinding),
                });
            }
        }

        function appendActionGroupBy(parentId: UniqueId) {
            const items: IMenuItemProps[] = [];
            return new (class {
                with = (ctor: { ID: string }) => {
                    const keybinding = molecule.action.queryGlobalKeybinding(ctor.ID);

                    items.push({
                        id: ctor.ID,
                        name: molecule.locale.localize(ctor.ID, ctor.ID),
                        keybinding: keybinding
                            ? molecule.action.convertSimpleKeybindingToString(keybinding)
                            : undefined,
                    });
                    return this;
                };
                exhaust = () => {
                    const menuItems = concatMenu(molecule.menuBar.get(parentId)?.children || [], items);
                    molecule.menuBar.update(parentId, () => ({
                        children: menuItems,
                    }));
                };
            })();
        }
    },
};
