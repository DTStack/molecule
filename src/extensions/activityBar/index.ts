import { Alignment, IExtension } from 'mo/types';
import { classify, concatMenu, sortByIndex } from 'mo/utils';

import { createContextMenu, createMenuDuplicate } from '../utils';

export const ExtendsActivityBar: IExtension = {
    id: 'ExtendsActivityBar',
    name: 'Extend The Default ActivityBar',
    activate: function (molecule): void {
        molecule.activityBar.onClick((item) => {
            if (item.id === molecule.activityBar.getCurrent()) {
                molecule.layout.setSidebar((prev) => !prev);
            } else {
                molecule.layout.setSidebar(true);
                molecule.activityBar.setCurrent(item.id);
                molecule.sidebar.setCurrent(item.id);
            }
        });

        molecule.activityBar.onContextMenu((pos, activityItem) => {
            const { ACTIVITYBAR_CONTEXTMENU = [] } = molecule.builtin.getModules();
            const activityBar = molecule.activityBar.getState().data;
            const [top = [], bottom = []] = classify(
                activityBar,
                (item) => item.alignment === Alignment.top
            );
            let contextMenu = concatMenu(
                createContextMenu(top.sort(sortByIndex)),
                createContextMenu(bottom.sort(sortByIndex)),
                ACTIVITYBAR_CONTEXTMENU
            );
            const target = contextMenu.find((item) => item.id === activityItem?.id);
            if (target) {
                contextMenu = concatMenu([createMenuDuplicate(target)], contextMenu);
            }
            molecule.contextMenu.open(
                contextMenu,
                pos,
                // remark current contextMenu
                molecule.builtin.getConstants().CONTEXTMENU_ITEM_ACTIVITYBAR
            );
        });
    },
};
