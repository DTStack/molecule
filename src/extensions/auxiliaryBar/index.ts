import { IExtension } from 'mo/types';

export const ExtendsAuxiliaryBar: IExtension = {
    id: 'ExtendsAuxiliaryBar',
    name: 'Extend The Default AuxiliaryBar',
    activate: function (molecule): void {
        molecule.auxiliaryBar.onTabClick((id) => {
            molecule.auxiliaryBar.toggle(id);
        });
    },
};
