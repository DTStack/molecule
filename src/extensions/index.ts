import { ExtendsActions } from './actions';
import { ExtendsActivityBar } from './activityBar';
import { ExtendsAuxiliaryBar } from './auxiliaryBar';
import { ExtendsContextMenu } from './contextMenu';
import { ExtendsEditor } from './editor';
import { ExtendsEditorTree } from './editorTree';
import { ExtendsExplorer } from './explorer';
import { ExtendsFolderTree } from './folderTree';
import { LocalesExtends } from './locales';
import { ExtendsMenuBar } from './menuBar';
import { ExtendsNotification } from './notification';
import { ExtendsPanel } from './panel';
import { ExtendsSearch } from './search';
import { ExtendsSettings } from './settings';
import { ExtendsSidebar } from './sidebar';
import { ExtendsStatusBar } from './statusBar';
import { ColorThemeExtension } from './themes';

export default [
    LocalesExtends,
    ColorThemeExtension,
    ExtendsActions,
    ExtendsEditor,
    ExtendsActivityBar,
    ExtendsEditorTree,
    ExtendsPanel,
    ExtendsMenuBar,
    ExtendsNotification,
    ExtendsStatusBar,
    ExtendsSearch,
    ExtendsSettings,
    ExtendsFolderTree,
    ExtendsSidebar,
    ExtendsExplorer,
    ExtendsContextMenu,
    ExtendsAuxiliaryBar,
];
