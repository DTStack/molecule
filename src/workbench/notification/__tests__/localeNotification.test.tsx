import 'reflect-metadata';
import React from 'react';
import { create } from 'react-test-renderer';
import { container } from 'tsyringe';
import { LocaleService } from 'mo/i18n';
import { ExtendsLocales } from 'mo/extensions/locales-defaults';
import { fireEvent, render } from '@testing-library/react';
import LocaleNotification from '../notificationPane/localeNotification';
import '@testing-library/jest-dom';

describe('The LocaleNotification Component', () => {
    // initial locales
    const localeService = container.resolve(LocaleService);
    localeService.initialize(ExtendsLocales.contributes!.languages!, 'en');

    test('Match Snapshot', () => {
        const component = create(<LocaleNotification locale="chinese" />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Should support to reload via button', () => {
        const originalFunction = window.location.reload;
        const mockFn = jest.fn();
        Reflect.deleteProperty(window, 'location');
        Object.defineProperty(window, 'location', {
            writable: true,
            value: { reload: mockFn },
        });
        const { getByText } = render(<LocaleNotification locale="chinese" />);

        fireEvent.click(getByText('Confirm Reload'));

        expect(jest.isMockFunction(window.location.reload)).toBeTruthy();
        expect(mockFn).toBeCalled();

        window.location.reload = originalFunction;
    });

    test('Should support to reload by pressing the Enter key.', async () => {
        const originalFunction = window.location.reload;
        const mockFn = jest.fn();
        Reflect.deleteProperty(window, 'location');
        Object.defineProperty(window, 'location', {
            writable: true,
            value: { reload: mockFn },
        });
        const { container } = render(<LocaleNotification locale="zh-CN" />);
        const elem = container.querySelector('button')!;

        expect(elem).toBeInTheDocument();
        expect(elem === document.activeElement).toBeTruthy();

        fireEvent.click(elem);
        expect(mockFn).toBeCalled();
        window.location.reload = originalFunction;
    });
});
