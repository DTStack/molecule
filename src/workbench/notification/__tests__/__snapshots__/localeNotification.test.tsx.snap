// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The LocaleNotification Component Match Snapshot 1`] = `
<div
  style={
    Object {
      "lineHeight": "1.5",
    }
  }
>
  <div
    style={
      Object {
        "direction": "ltr",
        "textAlign": "left",
        "whiteSpace": "normal",
        "width": 440,
      }
    }
  >
    <p
      style={
        Object {
          "fontWeight": "bold",
        }
      }
    >
      A reload is required for the change in display language to take effect.
    </p>
    <p>
      Press the confirm button to reload and change the display language.
    </p>
  </div>
  <div
    style={
      Object {
        "marginBottom": 2,
      }
    }
  >
    <button
      className="mo-btn mo-btn--normal"
      onClick={[Function]}
      style={
        Object {
          "margin": "0 0 0 auto",
          "width": 150,
        }
      }
    >
      Confirm Reload
    </button>
  </div>
</div>
`;
